Add an option to make use of GNU Readline.

ethorondil opened the pull request:
GNU readline has a lot of features, so it's nice to use the library
directly rather than try to reimplement it -- inevitably, some
features would be missing.

This patch works with Readline 7.0 on my linux machine.  I don't know
whether it works on other platforms (libreadline should be available in
brew and mingw, though.)

Known outstanding issues:

- Spellcheck is a bit wonky with punctuation, but that's really an
  existing bug in recheckWholeLine() that can be addressed separately.

- You can't create multiline inputs, and interaction becomes slow for
  very large (>100 word) lines because of spellcheck.

- Readline doesn't behave all that intuitively with a mouse.

- The build only uses -lreadline on unix, not mac or windows.  I have
  no ability to test other platforms.

Only the last one seems serious to me.

Another possible issue is that I reorganized the code a bit even for
when the readline option is disabled.  The only way in which I know
this changes behavior is that now potential user-defined key bindings
take priority over commandLine bindings; in my opinion that is better
behavior, and it made the code simpler.  But it's certainly possible
that I accidentally changed the behavior in other ways -- I don't know
the expected behavior well enough to test it thoroughly.

#### Other info (issues closed, discussion etc)

Discussed in issue #2650.

vadi2 commented:
Cool!

I'm not familiar with GNU Readline except that it can provide tab completion for program arguments... is it  basically adding the functions on https://en.wikipedia.org/wiki/GNU_Readline ?

ethorondil commented:
Pretty much.  There are some other functions, as described in https://tiswww.cwru.edu/php/chet/readline/rluserman.html , but the wikipedia page mentions the most important ones.

SlySven commented on src/TCommandLine.cpp:
```diff
@@ -115,85 +193,198 @@ bool TCommandLine::event(QEvent* event)
+        int modifier = 1 + 1 * (bool(ke->modifiers() & Qt::ShiftModifier)) + 2 * (bool(ke->modifiers() & Qt::AltModifier)) + 4 * (bool(ke->modifiers() & Qt::ControlModifier));
+        QString specialPrefix = QString("\x1b[");
+        if (modifier > 1)
+            specialPrefix = specialPrefix+QString("1;")+('0'+modifier);
```
Multiple concatenations are less than idea, perhaps:
```
    specialPrefix.append(QStringLiteral("1;%1").arg(QString::toNumber(modifier)))
```
would work more efficiently?

SlySven commented on src/TCommandLine.cpp:
```diff
@@ -115,85 +193,198 @@ bool TCommandLine::event(QEvent* event)
+        case Qt::Key_D:
+            // Readline treats ^D on empty line as EOF and quits.  So don't send it.
+            if ((ke->modifiers() == Qt::ControlModifier) && rl_end == 0)
+                keyText = QString("");
```
:book: You do not need to assign an empty string to clear a `QString`:
```
keyText.clear();
```
imparts the same intent and avoids another raw-string to `QString` creation {anyhow the result - an empty `QString` is almost indistinguishable from a null `QString` - only `QString::isNull()` can tell the difference so the recommendation when you *do* want to do an assignment (e.g. in a constructor's initialiser list) is to use `QString()`, though as the default constructor for the QString is the null one anyhow it isn't *needed* in that case}.

SlySven commented on src/TCommandLine.cpp:
```diff
@@ -115,85 +193,198 @@ bool TCommandLine::event(QEvent* event)
+        // Missing: C-x for cut, tab completion using GNU readline
+
+        if (mpHost->mEnableReadline) {
+            if (keyText.length() == 0)
```
```
            if (keyText.isEmpty())
```

SlySven commented on src/TCommandLine.cpp:
```diff
@@ -115,85 +193,198 @@ bool TCommandLine::event(QEvent* event)
+            if (keyText.length() == 0)
+                return false;
+            if (keyText.length() == 1 && (ke->modifiers() & Qt::AltModifier)){
+                keyText = "\x1b" + keyText;
```
```
                keyText.prepend(QLatin1Char('\x1b'));
```

SlySven commented on src/TCommandLine.cpp:
```diff
@@ -115,85 +193,198 @@ bool TCommandLine::event(QEvent* event)
+            // Copy the widget's view to readline's view.
+
+            // In order to figure out where the current cursor position is
+            // in the UTF-8 encoded string, we separately encode the
```
The widget's text is NOT *UTF-8* encoded - it is *UTF-**16*** - does that break things here?

ethorondil replied:
See my longer comment: this is referring to the encoding _for readline_.  We want to convert the QTextCursor position (which is a number of characters) into the rl_line_buffer (which is a number of bytes in the often-but-not-necessarily-utf8 string).

vadi2 commented:
OK, let's get the behaviour ironed out on Linux - and then I'll help with the macOS and Windows versions. I agree the huge lag when you enter text has to go away + mouse handling is really broken in the widget with readline.

ethorondil commented:
> What happens to our existing history code and the data stored in `mHistoryList`?

In the updated, `mHistoryList` is maintained as well.

> I do have some concerns that this will break existing functionality - though I am not sure that I have identified all of the existing stuff.
> 
> A further complication is that the library is a C one which I think nowadays works with UTF-8 encoded text - however `QString`s use a different _**variable**-length-encoding_ i.e. UTF-16 so the indexing of the graphemes (what the user perceives as _characters_) between those two is going to be really evil...

It is a little annoying, but it's not that bad: since we only need to keep track of a single cursor, we can just decode (the bytes left of the cursor) and (the bytes right of the cursor) separately to convert the C library's byte index into a character index.  QString's encoding doesn't come up.

The library uses the current locale to decide on the encoding.  (This is why one line of the commit changes a lua call from `setlocale('C')` to `setlocale('C', 'numeric')`:  if the LC_CTYPE is C, then Unicode can't be encoded at all.)  I was mistakenly always encoding with toUtf8 rather than toLocal8Bit; this has been fixed.

ethorondil commented:
> OK, let's get the behaviour ironed out on Linux - and then I'll help with the macOS and Windows versions. I agree the huge lag when you enter text has to go away + mouse handling is really broken in the widget with readline.

What huge lag are you talking about?  I don't notice any lag until entering >100 words in a single line, which seems like a pretty strange case.  (That lag, I suppose, could be 

As for mouse handling, well, most readline applications completely ignore the mouse.  So it's not actually clear to me that (for example) DEL when stuff is selected _should_ delete the selected text -- it doesn't do so in other readline applications, like bash or the lua REPL in xterm or gnome-terminal.

SlySven commented:
> The library uses the current locale to decide on the encoding. (This is why one line of the commit changes a lua call from setlocale('C') to setlocale('C', 'numeric'): if the LC_CTYPE is C, then Unicode can't be encoded at all.) I was mistakenly always encoding with toUtf8 rather than toLocal8Bit; this has been fixed.

Does that `setlocale` work on just the readline library code or does it act on the whole application - because if it is the latter then I think it will break the rest of Mudlet for non en_US users - I mean I run with my locale set to `en_GB` or sometimes `fr_FR` so how does it behave if I enter accented text I wonder...

I'll give it a spin this morning and see!

vadi2 commented:
Ah ok, so we'll need to give people a warning that mouse handling is expected to be different.

> What huge lag are you talking about? I don't notice any lag until entering >100 words in a single line, which seems like a pretty strange case.

Definitely not - people can write entire essays in the game! Have you profiles this, where is the lag coming from?

ethorondil commented:
> Definitely not - people can write entire essays in the game! Have you profiles this, where is the lag coming from?

It comes from spell check.  At the moment, whenever a character is pressed, it rechecks the whole line.  Avoiding that is... kind of annoying, since it's not easy to tell where the changes made by readline appear (for example, undo (C-\) could make a large non-local change).  I suppose one could cache the results of spellCheck() to make checking the whole line faster, maybe?

ethorondil commented:
> > The library uses the current locale to decide on the encoding. (This is why one line of the commit changes a lua call from setlocale('C') to setlocale('C', 'numeric'): if the LC_CTYPE is C, then Unicode can't be encoded at all.) I was mistakenly always encoding with toUtf8 rather than toLocal8Bit; this has been fixed.
> 
> Does that `setlocale` work on just the readline library code or does it act on the whole application - because if it is the latter then I think it will break the rest of Mudlet for non en_US users - I mean I run with my locale set to `en_GB` or sometimes `fr_FR` so how does it behave if I enter accented text I wonder...

Maybe I wasn't clear -- I'm not using setlocale() in the readline code.  I removed (or rather, limited) a setlocale() elsewhere in the Mudlet code.  It might be good to add a `setlocale(LC_CTYPE, "UTF-8")` call, however: LC_CTYPE can only be C or UTF-8, and if it's C then non-ascii characters will behave quite poorly.  But the user's environment probably already has it set to UTF-8, and if it's not then other readline applications don't handle wide characters either, so the choice isn't 100% clear (maybe the user had it set otherwise for a reason?).

vadi2 commented:
Try the caching. We can't explain away such lag, it's a bad user experience. You'd hate it if you downloaded a program that had this lag :(

SlySven commented:
Looking at the action of other locales has, for me so far, been inconclusive but I have not been that focused on this PR recently...

ethorondil commented:
OK, I:
 - added the `setlocale(LC_CTYPE, "UTF-8")` call just in case.
 - optimized the spellcheck.

Unfortunately, the spellcheck optimization isn't quite what I would like.  The bottleneck is actually iterating over each word and changing its `QTextCharFormat`, not figuring out what that format should be; it seems like that iteration is quadratic time.  As far as I can tell, there is no fast way to build a `QPlainTextEdit` document from a list of words of known formats -- repeated calls to `insertPlainText` is similarly slow.

The "solution" I implemented is to make it so it only needs to change `QTextCharFormat` for each word that is incorrectly spelled.  So as long as the essays being written have relatively few spelling errors, it works OK.

vadi2 commented:
Awesome, going to look at it now.

The next release is this weekend and feature freeze is tomorrow, so this **might** not make it into 3.22 as I've got a lot on, but the next release will be just a month away.

vadi2 commented:
Not something necessary for this PR - but readline is clearly oriented a terminal-like input line, not a standard input line. What do you think about simulating (in readline mode) the rectangle cursor? Afaik you can paint a custom one with https://doc.qt.io/qt-5/qtextlayout.html#drawCursor.

vadi2 commented:
Newlines seem to get converted to tabs when pasted, is that intended?

vadi2 commented:
`Ctrl+L` doesn't clear the screen as Wikipedia suggests.

vadi2 commented:
I've played around with it more and it seems really cool and nice. But what's the practical usecase for it? The input line isn't a text editor, it has a different purpose. We'd still put it in but I wonder

Kebap commented:
Following up on that argument, Mudlet does have big text editor windows for scripts, etc.
They use completely different shortcuts however. Do we really want users to learn two sets?
Not sure which shortcuts and features @ethorondil originally missed to propose this issue.

ethorondil commented:
> The input line isn't a text editor, it has a different purpose.

It seems to me that the input line's purpose is pretty similar to the
command line on a terminal, which is what readline was designed for.
Most of the time you're writing simple commands, but occasionally
longer ones, and you refer to older commands with some frequency.

> Not sure which shortcuts and features @ethorondil originally missed to propose this issue.

A few things.  First, there are minor shortcuts that are just finger
macros from years of usage of the command line: my fingers expect
`C-a`, `C-w`, `C-y` do go to beginning of line, delete the previous
word, and paste the deleted word.  Any of those could be done with a
couple key chords in QTextEdit, but my muscle memory gets confused.
(Also, the flexibility in word boundaries between `C-w` and `M-DEL` is
useful and can't be easily replicated.)

More substantively, there's usage of command line history search.  For
instance, suppose I write a complicated line like

  `unlock chest;open chest;get bar chest`

that I might want to run more than once.  If I'm going to run that
chain many times, it makes sense to write an alias for it.  But if I'm
only going to do it a few times, I can write it once then invoke it
later with something like

  `C-rbar<enter>`

Moreover, unlike with an alias, I don't need to plan ahead: as soon as
I realize I need it again, I can go back and find it and fine-tune it.

vadi2 commented:
Will digest this, but just dropping a note - [hunspell 1.7.0](https://github.com/hunspell/hunspell/releases/tag/v1.7.0) mentions massive performance improvements for mispellings.

vadi2 commented:
Sorry, can you explain how does command line history search work? (in terms of `Ctrl`, `Alt`, etc)

keneanung commented on cmake/FindREADLINE.cmake:
```diff
@@ -0,0 +1,47 @@
+# - Try to find readline include dirs and libraries 
```
I'm not quite happy with this find module as it is exporting variables and such. Modern find modules should export targets with all needed variables (see https://pabloariasal.github.io/2018/02/19/its-time-to-do-cmake-right/). I have a branch where I started modernizing existing find modules.

To avoid being a show-stopper I'm okay with this form for now, I just wanted to create a bit sensitivity for that topic :wink:

vadi2 replied:
Good to know!

vadi2 commented:
@ethorondil ping - does the command line history search with this implementation, do you know?

vadi2 commented:
There hasn't been much interest in this cool, but advanced command line editor - closing for now.
